### DISCUSSION TITLE
Database optimization

### DISCUSSION THREAD

**Discussion by smichel17** [streetcomplete/StreetComplete#3603] [Ideas]
Have we done any optimization of the database queries or table layout? We've bumped into performance constraints a couple times, most recently with #3594. I know some time recently-ish there was a change where a bunch of queries were grouped into one. Have we done any further performance analysis? I don't necessarily mean *implementing* all the worthwhile optimizations, but just figuring out where we have room to easily improve performance, and where improved performance would require additional complexity, like adding layers of caching.

**Comment by Helium314** [comment]
From reading some more, a spatial index could increase performance considerably.
But it looks like even testing this is quite a bit of work, as in my understanding it requires using a different database. At least an sqlite build with r-tree, or spatialite for further enhancements.

**Reply by smichel17** [reply]
> a spatial index could increase performance considerably

How much is considerably? Also, any idea how much extra disk space a spatial index adds? I think the current db implementation is provided by Android, so adding a (presumably) 3rd party dependency would be a significant step.

**Reply by Helium314** [reply]
> How much is considerably?

I would tell you if I could. From information I found, the gain is larger if you are interested in a small area compared to your whole database.
But it also depends very much on the exact query, some comparison I found: https://www.researchgate.net/publication/316278667_PERFORMANCE_COMPARISON_OF_SPATIAL_INDEXING_STRUCTURES_FOR_DIFFERENT_QUERY_TYPES and https://automating-gis-processes.github.io/site/notebooks/L3/spatial_index.html
Summary: R-tree is somewhere between same speed and 15 times faster than no spatial index (for the used dataset).

For determining the actual perfornance gain, I think a good way to find out would be running some queries commonly used by SC on databases with and without spatial index.

**Reply by Helium314** [reply]
Hmm, looks like sqlite has an R\*tree, and not R-tree: https://sqlite.org/rtree.html
So some queries are even much faster, but some queries are slower than without spatial index. And creation of the R*tree index might take long...

**Reply by Helium314** [reply]
One short test using [sqlite browser](https://sqlitebrowser.org) on my laptop, with a 78 MB database file from SC that has some 233000 entries in elements_geometry.
Create table with element_id, min_lat, max_lat, min_lon, max_lon and fill with data from elements_geometry: 173 ms normal, 3205 ms with R-tree
Query ca 50 x 150 m bounding box: 111 ms normal, 33 ms with R-tree
Query ca 50 x 30 m bounding box: 95 ms normal, 16 ms with R-tree

**Reply by smichel17** [reply]
How large is the DB after creating the table (ie. how big is the new table)?

edit: Oh, I think this might be a silly question— R-Tree is a different data structure for storing the info, not an additional index. Is that right?

**Reply by Helium314** [reply]
There is still some difference: after duplicating the data without R-tree: 89.2 MB, with R-tree: 90.6 MB. So R-tree adds some 15% (10.2 MB vs 11.6 MB)

But so far I failed in building SC with an sqlite library with R-tree.

**Reply by Helium314** [reply]
Well... the trick was doing a clean build.
So with an R-tree index for the elements_geometry table, element highlighting is definitely faster. But that's only https://github.com/streetcomplete/StreetComplete/blob/583da6031f3b3c2ee086c7f5c9ede08c5ee283da/app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataController.kt#L151 taking 20-50 ms instead of 100-200 ms.
The rest still takes long enough to reduce usefulness of this particular change: R-tree reduces overall time for `getMapDataWithGeometry` from 300-400 ms to 200-300 ms.

Time for downloading data and creating quests wasn't affected at all to my surprise. But there are more tables that could be changed...

**Reply by smichel17** [reply]
Do you think this is promising enough to create an issue about it (so this discussion won't be lost)?

/cc @westnordost

**Reply by westnordost** [reply]
yes

**Reply by Helium314** [reply]
I uploaded the changes I did for the test above: https://github.com/Helium314/StreetComplete/tree/rtree

Using the R-tree is [pretty simple](https://sqlite.org/rtree.html#using_the_r_tree_module), just table creation is a bit different:
```
CREATE VIRTUAL TABLE $NAME USING rtree(
            ${Columns.ELEMENT_ID},
            ${Columns.MIN_LATITUDE},
            ${Columns.MAX_LATITUDE},
            ${Columns.MIN_LONGITUDE},
            ${Columns.MAX_LONGITUDE},
            +${Columns.ELEMENT_TYPE} varchar(255) NOT NULL,
            +${Columns.GEOMETRY_POLYLINES} blob
```
First column is the primary index (int64), followed by min and max columns (float) for each dimension.
Further columns are added as usual, just with a prefixed `+`

And yes, further discussion should be done in an issue...

**Reply by smichel17** [reply]
I opened https://github.com/streetcomplete/StreetComplete/issues/3609

**Comment by smichel17** [comment]
> But it looks like even testing this is quite a bit of work

I was thinking more along the lines of "Find someone who knows sql(lite) well; ask them to glance over our schema/queries and let us know what they see that's likely slower than it could be." :joy:

**Comment by smichel17** [comment]
https://github.com/streetcomplete/StreetComplete/discussions/3603#discussioncomment-1871558 is pretty good data, thanks @Helium314.

Next step seems to be to analyze SC logs (adding additional logging if needed) to see how often these types of queries happen / whether the performance hit for insertion would be worth the speedup in the lookups.

**Reply by Helium314** [reply]
I think there is a lot that could be done... for example in the short test I did, I guess the slow part is getting the elements by the ids returned from the first query.
Since all the elements of the queried ids are close together, also the way/node/relation tables would profit from an R-tree index. And maybe instead of adding R-tree-indices to more tables, some other changes are possible. Like putting data of the elements in the same table as the geometry. However, I suspect there is a good reason why these tables are separate.

And there is the question of what library to use. [Spatialite](https://github.com/sevar83/android-spatialite) has some more interesting capabilities, but the Android versions I found don't look well maintained. For my test I used [this](https://github.com/requery/sqlite-android), which adds another 5 MB to SC.

**Reply by westnordost** [reply]
What about the support library? https://developer.android.com/reference/androidx/sqlite/db/package-summary

**Reply by Helium314** [reply]
I simply did not know about that.
I wanted to have some sqlite version with R-tree enabled (least work in case results show no improvement), and this wasn't part of my search results.

**Comment by westnordost** [comment]
There is a stale branch called `fetch-by-bbox`, maybe someone is interested in doing some actual performance comparisons. It:
- removes the index from element geometry table as well as the minlat, minlon, maxlat, maxlon
- adds index to node table on lat,lon
- adds index to ways table and relations table on (node) ids contained
- query by bbox works like: 1. get all nodes in given bbox, 2. get all ways that contain any of these nodes, 3. get all relations that contain any of these nodes or ways. Then, get all element geometry of the returned elements.

I am not entirely sure why I started this branch, it could be because
1. it reflects more closely what the OSM API returns on fetching a bbox
2. and/or because it **might** be faster because a) index or not, I imagine a query that looks for intersecting bboxes (fetch bounding box and bounding boxes of all the element geometry) is more expensive than a query that queries all points that are contained in a bounding box - i.e. should be half as many `>=` etc. comparison checks. b) because in places in the code where no element geometry is needed (is there any?), it is not necessary to query the element geometry table at all.

But given that it is stale, maybe I already did performance tests and was not content with the results. I do not remember.

**Reply by Helium314** [reply]
> I am not entirely sure why I started this branch

I think it was an attempt to fix slowdowns when changing to local OSM data, see https://github.com/streetcomplete/StreetComplete/issues/2803#issuecomment-830445155
